File: android/benchmarkable/src/main/java/com/example/platformchannelperformance/measuring/FpsMonitor.kt
package com.example.platformchannelperformance.measuring

import android.util.Log
import android.view.Choreographer
import android.view.Choreographer.FrameCallback
import android.view.Choreographer.getInstance
import com.example.benchmarkable.BuildConfig
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel

// https://github.com/wasabeef/Takt/blob/master/takt/src/main/java/jp/wasabeef/takt/Metronome.java
object FpsMonitor : FrameCallback {
    private val choreographer: Choreographer by lazy { getInstance() }
    private var lastFrameMs = 0.0
    private var experiment: String? = null


    fun initFlutterChannel(binaryMessenger: BinaryMessenger) {
        MethodChannel(binaryMessenger, "FpsEvents").setMethodCallHandler { call, result ->
            when (call.method) {
                "start" -> when (val name = call.arguments as? String) {
                    null -> result.error("bad args", null, null)
                    else -> {
                        start(name)
                        result.success(null)
                    }
                }

                "stop" -> {
                    stop()
                    result.success(null)
                }

                else -> result.notImplemented()
            }
        }
    }

    fun log(measurement: Double) {
        if (experiment == null) return
        Log.i("measure", "screenUpdate_$experiment: $measurement")
    }

    fun start(experimentName: String) {
        if (BuildConfig.DEBUG) return
        experiment = experimentName
        choreographer.postFrameCallback(this)
    }

    fun stop() {
        if (BuildConfig.DEBUG) return
        choreographer.removeFrameCallback(this)
        lastFrameMs = 0.0
        experiment = null
    }

    override fun doFrame(frameTimeNanos: Long) {
        val frameTimeMillis: Double = frameTimeNanos / 1000.0 / 1000.0
        if (lastFrameMs == 0.0) {
            lastFrameMs = frameTimeMillis
            choreographer.postFrameCallback(this)
            return
        }

        val msBetweenFrames = frameTimeMillis - lastFrameMs
        log(msBetweenFrames)
        lastFrameMs = frameTimeMillis
        choreographer.postFrameCallback(this)
    }
}

File: ios/PlatformChannelPerformance/util/FpsMonitor.swift
//
//  FpsMonitor.swift
//  PlatformChannelPerformance
//
//  Created by Markus Reinhold on 07.11.23.
//

import Foundation

import SwiftUI
import Metal
import QuartzCore
import os

class FpsMonitor: NSObject {
    private static var displayLink: CADisplayLink?
    private static var experiment: String?
    private static var lastFrameMs: CFAbsoluteTime = 0.0
    
    static var logCategory = OSLog(subsystem: "test.PlatformChannelPerformance", category: "measure")
    
    static func initFlutterChannel(flutterBinaryMessenger: FlutterBinaryMessenger) {
        let channel = FlutterMethodChannel(name: "FpsEvents", binaryMessenger: flutterBinaryMessenger)
        
        channel.setMethodCallHandler { (call: FlutterMethodCall, result: FlutterResult) -> Void in
            switch call.method {
            case "start":
                if let name = call.arguments as? String {
                    start(experimentName: name)
                    result(nil)
                } else {
                    result(FlutterError(code: "bad args", message: nil, details: nil))
                }
            case "stop":
                stop()
                result(nil)
            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }
    
    static func log(_ log: Double) {
        os_log("screenUpdate_%@: %f", log: logCategory, type: .debug, experiment ?? "", log)
    }
    
    static func start(experimentName: String) {
        displayLink = CADisplayLink(target: self, selector: #selector(updateFrameTime))
        displayLink?.add(to: .current, forMode: .common)
        experiment = experimentName
    }
    
    static func stop() {
        displayLink?.invalidate()
        displayLink = nil
        lastFrameMs = 0.0
    }
    
    @objc static private func updateFrameTime(displayLink: CADisplayLink) {
        let currentFrameMs = displayLink.timestamp * 1000
        if (lastFrameMs == 0.0) {
            lastFrameMs = currentFrameMs
            return
        }
        let msBetweenFrames = (currentFrameMs - lastFrameMs)
        log(msBetweenFrames)
        lastFrameMs = currentFrameMs
    }
}
